### PR TITLE
Feat: Fishing

### PR DESCRIPTION
**Pull request by Zintixx**
## Fishing
Mainly just implementation of fishing

 - Parsing `FishMetadata`, `FishingSpotMetadata`, and `FishingRodMetadata`
 - Added Type and Attribute to `MapMetadata` in order to figure out if a block is a (true) liquid
 - Adjusted `Block.GetClosestBlock` to what was discussed on Discord
 - Added `Direction`
	 - The values here and directions here are labeled as how the game _looks_, rather than taking into consideration the 45 degree map rotation the game has baked in (for majority of the maps). Totally up for discussion it needs to be changed.
- Partial implementation of `GuideObject`. It's currently missing syncing to other players. WIP!

### Fishing Rates
These are the rates I'm working with. Most are guestimates.

- Fishing Success Rate - 90%
- Mini Game Chance - 10%
 - Fish Rarity Catch
	 - Common - 60%
	 - Rare - 15%
	 - Exceptional - 10%
	 - Epic - 5%
- Fish Size
	- Small - 85%
	- Medium - 8%
	- Large/Prized - 3% (this one I know is for sure after cross checking my KMS2 fishing album)
- Item Catch Rate - 10%
	- Items are split into 4 groups:
		- Trash	- 82.5%
		- Light Boxes - 15%
		- Heavy Boxes - 2%
		- Skins/Cosmetics - 0.5%
- Mastery increase for existing fish - 10%

### CODE REVIEW

File: MapleServer2/Data/Static/MapMetadataStorage.cs [hunk 1]
```diff
@@ -28,7 +29,31 @@ public static MapMetadata GetMetadata(int mapId)
...
+        public static bool BlockAboveExists(int mapId, CoordS coord)
+        {
+            MapMetadata mapD = GetMetadata(mapId);
+            coord.Z += Block.BLOCK_SIZE;
+            MapBlock block = mapD.Blocks.FirstOrDefault(x => x.Coord == coord);
+            if (block == null)
+            {
+                return false;
+            }
+            return true;
```
Comment (AlanMorel) [inline MapleServer2/Data/Static/MapMetadataStorage.cs:50]: Can be shorted to just ```return block != null;```

File: MapleServer2/PacketHandlers/Game/FishingHandler.cs [hunk 2]
```diff
@@ -1,47 +1,502 @@
...
+                        checkBlock.Z = startCoord.Z; // reset Z
+                    }
+                    checkBlock.X += Block.BLOCK_SIZE;
+                    checkBlock.Y = startCoord.Y; // reset Y
+                }
+            }
+            return blocks;
+        }
+
+        private static bool IsLiquidBlock(MapBlock block)
```
Comment (AlanMorel) [inline MapleServer2/PacketHandlers/Game/FishingHandler.cs:276]: This seems better off living inside MapBlock to allow it to be reusable in other situations

**Review by AlanMorel — CHANGES_REQUESTED** [review]

File: MapleServer2/PacketHandlers/Game/FishingHandler.cs (hunk 2 above)
Comment (Zintixx) [reply MapleServer2/PacketHandlers/Game/FishingHandler.cs:276]: I don't think we usually keep methods inside metadata information. Not opposed to it, but just want to confirm

File: MapleServer2/PacketHandlers/Game/FishingHandler.cs [hunk 3]
```diff
@@ -1,47 +1,502 @@
...
+                block.Attribute == "seawater" ||
+                block.Attribute == "devilwater" ||
+                block.Attribute == "lava" ||
+                block.Attribute == "poison" ||
+                block.Attribute == "oil" ||
+                block.Attribute == "emeraldwater")
+            {
+                return true;
+            }
+            return false;
```
Comment (angelotadeucci) [inline MapleServer2/PacketHandlers/Game/FishingHandler.cs:293]:
```suggestion
            return block.Attribute == "water" ||
                block.Attribute == "seawater" ||
                block.Attribute == "devilwater" ||
                block.Attribute == "lava" ||
                block.Attribute == "poison" ||
                block.Attribute == "oil" ||
                block.Attribute == "emeraldwater";
```

File: MapleServer2/PacketHandlers/Game/FishingHandler.cs [hunk 4]
```diff
@@ -1,47 +1,502 @@
...
+                                continue;
+                            }
+
+                            blocks.Add(block);
+                            checkBlock.Z -= Block.BLOCK_SIZE;
+                        }
+                        checkBlock.Z = startCoord.Z; // reset Z
+                    }
+                    checkBlock.X += Block.BLOCK_SIZE;
+                    checkBlock.Y = startCoord.Y; // reset Y
```
Comment (angelotadeucci) [inline MapleServer2/PacketHandlers/Game/FishingHandler.cs:270]: Could put this inside a function so it gets reused

File: MapleServer2/PacketHandlers/Game/FishingHandler.cs [hunk 5]
```diff
@@ -1,47 +1,502 @@
...
+
+            //determine fish size
+            int fishSize = rnd.NextDouble() switch
             {
-                case 0:
-                    long rodItemUid = packet.ReadLong(); // Testing RodItemId: 32000055
-                    PacketWriter pWriter = PacketWriter.Of(SendOp.FISHING);
-                    pWriter.WriteHexString("04 00 08 00 00 00 0C F7 10 00 81 96 98 00 19 00 00 00 98 3A 00 00 01 00 0B F8 10 00 81 96 98 00 19 00 00 00 98 3A 00 00 01 00 0E F6 10 00 81 96 98 00 19 00 00 00 98 3A 00 00 01 00 0D F7 10 00 81 96 98 00 19 00 00 00 98 3A 00 00 01 00 0C F8 10 00 81 96 98 00 19 00 00 00 98 3A 00 00 01 00 0F F6 10 00 81 96 98 00 19 00 00 00 98 3A 00 00 01 00 0E F7 10 00 81 96 98 00 19 00 00 00 98 3A 00 00 01 00 0D F8 10 00 81 96 98 00 19 00 00 00 98 3A 00 00 01 00");
+                >= 0.0 and < 0.03 => rnd.Next(fish.SmallSize[0], fish.SmallSize[1]),
+                >= 0.03 and < 0.15 => rnd.Next((fish.SmallSize[1]), (fish.BigSize[0])),
```
Comment (angelotadeucci) [inline MapleServer2/PacketHandlers/Game/FishingHandler.cs:324]:
```suggestion
                >= 0.03 and < 0.15 => rnd.Next(fish.SmallSize[1], fish.BigSize[0]),
```

File: MapleServer2/PacketHandlers/Game/FishingHandler.cs [hunk 6]
```diff
@@ -1,47 +1,502 @@
...
+                }
+
+                session.Send(FishingPacket.CatchFish(session.Player, fish, fishSize, true));
+                HandleCatchItem(session);
+                session.Player.Levels.GainMasteryExp(MasteryType.Fishing, fish.Rarity);
+            }
+            else
+            {
+                session.Send(FishingPacket.CatchFish(session.Player, fish, fishSize, false));
+            }
```
Comment (angelotadeucci) [inline MapleServer2/PacketHandlers/Game/FishingHandler.cs:348]:
```suggestion
                session.Player.Levels.GainMasteryExp(MasteryType.Fishing, fish.Rarity);
                return;
            }

            session.Send(FishingPacket.CatchFish(session.Player, fish, fishSize, false));
```

File: MapleServer2/PacketHandlers/Game/FishingHandler.cs [hunk 7]
```diff
@@ -1,47 +1,502 @@
...
+                return;
+            }
 
-                    session.Send(pWriter);
-                    //session.Send(GuideObjectPacket.Bracket(session.FieldPlayer));
-                    session.Send(FishingPacket.Start(rodItemUid));
+            if (!IsLiquidBlock(block))
+            {
+                return;
+            }
```
Comment (angelotadeucci) [inline MapleServer2/PacketHandlers/Game/FishingHandler.cs:386]:
```
            if (block == null || !IsLiquidBlock(block))
            {
                return;
            }
```

File: MapleServer2/PacketHandlers/Game/FishingHandler.cs [hunk 8]
```diff
@@ -1,47 +1,502 @@
...
+                FishingRodMetadata rodMetadata = FishingRodMetadataStorage.GetMetadata(session.Player.FishingRod.Function.Id);
+                int rodFishingTick = fishingTick - rodMetadata.ReduceTime; // 10000 is the base tick for fishing
+                fishingTick = rnd.Next(rodFishingTick - (rodFishingTick / 3), rodFishingTick); // chance for early catch
+            }
+            else
+            {
+                fishingTick = 20000; // if tick is over the base fishing tick, it will fail
+            }
+
+            session.Send(FishingPacket.Start((session.ClientTick + fishingTick), minigame));
```
Comment (angelotadeucci) [inline MapleServer2/PacketHandlers/Game/FishingHandler.cs:412]:
```suggestion
            session.Send(FishingPacket.Start(session.ClientTick + fishingTick, minigame));
```

File: MapleServer2/PacketHandlers/Game/FishingHandler.cs [hunk 9]
```diff
@@ -1,47 +1,502 @@
...
+                        }
+                        checkBlock.Z = startCoord.Z; // reset Z
+                    }
+                    checkBlock.Y += Block.BLOCK_SIZE;
+                    checkBlock.X = startCoord.X; // reset X
+                }
+            }
+            else if (direction == Direction.SOUTH_EAST)
+            {
+                checkBlock.X -= (2 * Block.BLOCK_SIZE); // start at the corner
```
Comment (angelotadeucci) [inline MapleServer2/PacketHandlers/Game/FishingHandler.cs:242]: All others too
```suggestion
                checkBlock.X -= 2 * Block.BLOCK_SIZE; // start at the corner
```

File: MapleServer2/PacketHandlers/Game/FishingHandler.cs [hunk 10]
```diff
@@ -1,47 +1,502 @@
...
+            session.FieldManager.AddGuide(fieldGuide);
+
+            session.Send(FishingPacket.LoadFishTiles(fishingBlocks, rodMetadata.ReduceTime));
+            session.FieldManager.BroadcastPacket(GuideObjectPacket.Add(session.FieldPlayer));
+            session.Send(FishingPacket.PrepareFishing(fishingRodUid));
+        }
+
+        private static CoordF GetObjectBlock(List<MapBlock> blocks, CoordF playerCoord)
+        {
+            MapBlock guideBlock = blocks.OrderBy(o => Math.Sqrt(Math.Pow((playerCoord.X - o.Coord.X), 2) + Math.Pow((playerCoord.Y - o.Coord.Y), 2))).First();
```
Comment (angelotadeucci) [inline MapleServer2/PacketHandlers/Game/FishingHandler.cs:130]:
```suggestion
            MapBlock guideBlock = blocks.OrderBy(o => Math.Sqrt(Math.Pow(playerCoord.X - o.Coord.X, 2) + Math.Pow(playerCoord.Y - o.Coord.Y, 2))).First();
```

File: MapleServer2/Servers/Game/FieldManager.cs [hunk 11]
```diff
@@ -247,6 +247,21 @@ public void AddMob(IFieldObject<Mob> fieldMob)
...
+        }
+
+        public bool RemoveGuide(IFieldObject<GuideObject> fieldGuide)
+        {
+            if (!State.RemoveGuide(fieldGuide.ObjectId))
+            {
+                return false;
+            }
+
+            return true;
```
Comment (angelotadeucci) [inline MapleServer2/Servers/Game/FieldManager.cs:262]:
```suggestion
            return State.RemoveGuide(fieldGuide.ObjectId);
```

File: MapleServer2/Types/Fishing.cs [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+
+namespace MapleServer2.Types
+{
+    public class Fishing
+    {
+        public int FishId;
+        public int TotalCaught;
+        public int TotalPrizeFish;
+        public int LargestFish;
+        public Fishing() { }
```
Comment (angelotadeucci) [inline MapleServer2/Types/Fishing.cs:17]: Don't need to add the default constructor, `AddExistingFish()` and `AddNewFish()` can be static

**Review by angelotadeucci — APPROVED** [review]
Good work!

File: MapleServer2/PacketHandlers/Game/FishingHandler.cs (hunk 4 above)
Comment (Zintixx) [reply MapleServer2/PacketHandlers/Game/FishingHandler.cs:270]: I'm not exactly sure how. If the player faces a different direction, X and Y values swap and so does the operator from - to + (vice versa).

Comment (angelotadeucci) [reply MapleServer2/PacketHandlers/Game/FishingHandler.cs:270]: for for loop for Z is the same for all the four cases

File: MapleServer2/PacketHandlers/Game/FishingHandler.cs (hunk 2 above)
Comment (AlanMorel) [reply MapleServer2/PacketHandlers/Game/FishingHandler.cs:222]: Yeah good point, nevermind

**Review by AlanMorel — APPROVED** [review]

**Pull request merged**